tests/js/plugin/enhancedanchoring-test.coffee

```diff
@@ -0,0 +1,198 @@
+
+class FakeAnchor extends Annotator.Anchor
+
+  _getSegment: -> "html segment for " + @id
+
+  constructor: (manager, annotation, target) ->
+    super manager, annotation, target, 42, 42,
+      "fake quote for" + target.id
+
+    @id = "fake anchor for " + target.id
+
+describe 'Annotator.Plugin.EnhancedAnchoring', ->
+  sandbox = null
+
+  createAnchoringManager = ->
+    annotator =
+      publish: (event) ->
+
+    am = new Annotator.Plugin.EnhancedAnchoring()
+    am.annotator = annotator
+    am.pluginInit()
+
+    am._chooseAccessPolicy()
+
+    am.strategies.push
+      name: "dummy anchoring strategy"
+      code: (annotation, target) ->
+        new FakeAnchor am,annotation, target
+
+    am
+
+  createTestAnnotation = (id, targets = 1) ->
+    id: "annotation " + id
+    target: (("target " + id + "-" + num) for num in [1 .. targets])
+    anchors: []
+
+  beforeEach ->
+    sandbox = sinon.sandbox.create()
+    sandbox.stub Annotator.TextHighlight, 'createFrom',
+      (segment, anchor, page) -> {segment, anchor, page}
+
+  afterEach ->
+    sandbox.restore()
+
+  describe "createAnchor", ->
+
+    it 'adds an anchor property to the annotations', ->
+      am = createAnchoringManager()
+      ann1 = createTestAnnotation "a1", 2
+
+      anchor11 = am.createAnchor(ann1, ann1.target[0]).result
+      anchor12 = am.createAnchor(ann1, ann1.target[1]).result
+
+      assert.isArray ann1.anchors
+      assert.include ann1.anchors, anchor11
+      assert.include ann1.anchors, anchor12
+      assert.equal ann1.anchors.length, 2
+
+    it 'adds an annotation property to the created anchors', ->
+      am = createAnchoringManager()
+      ann1 = createTestAnnotation "a1"
+      anchor11 = am.createAnchor(ann1, ann1.target[0]).result
+      assert.equal anchor11.annotation, ann1
+
+    it 'adds a target property to the created anchors', ->
+      am = createAnchoringManager()
+      ann1 = createTestAnnotation "a1"
+      anchor11 = am.createAnchor(ann1, ann1.target[0]).result
+
+      assert.equal anchor11.target, ann1.target[0]
+
+    it 'creates the anchors from the right targets', ->
+      am = createAnchoringManager()
+      ann1 = createTestAnnotation "a1"
+      anchor11 = am.createAnchor(ann1, ann1.target[0]).result
+      assert.equal anchor11.id, "fake anchor for " + anchor11.target.id
+
+    it 'adds the created anchors to the correct per-page array', ->
+      am = createAnchoringManager()
+      ann1 = createTestAnnotation "a1"
+      anchor11 = am.createAnchor(ann1, ann1.target[0]).result
+      assert.include am.anchors[anchor11.startPage], anchor11
+
+    it 'adds the created highlights to the anchors', ->
+      am = createAnchoringManager()
+      ann1 = createTestAnnotation "a1"
+      anchor11 = am.createAnchor(ann1, ann1.target[0]).result
+
+      assert.isObject anchor11.highlight
+      page = anchor11.startPage
+      hl11 = anchor11.highlight[page]
+      assert.ok hl11
+      assert.equal hl11.page, page
```

Comment:
On 2015-02-05 18:38, Aron Carroll wrote:
>
> Is there a reason behind the numbers on the variables?
>

Only when we are dealing with several anchors, or several targets per
anchors.

(I'll remove them when this is not the case.)